### ISSUE TITLE
Clarification of project purpose/intent...

### ISSUE THREAD

**Issue by roskelleycj** [cdklabs/cdk-pipelines-github#147]
So this project appears to build on the concepts from CDK Pipelines.  At a high level the CDK Pipeline concepts are:
* Define a Pipeline that includes a 'Build' / 'Synth' step
* Define Stages that include 'Stacks' and their 'ENV' settings, adding order if needed.
* And of course align permissions needed to perform the deployments that will occur.

Again this is probably to high level, but it helps if I can point out that this statement is what confuses me:

```
This module (cdk-pipelines-github) serves the same surface area, except that it is implemented with GitHub Workflows.
```

I initially had envisioned that CodePipeline would be created/triggered when using GitHubWorkflow, but in fact that is not the case.  CodePipeline and CodeBuild are not created/used when GitHubWorkflow is being used.  The 'pipeline' is fully defined in GithubWorkflow and uses permissions to S3 and CFN to achieve the same effects as CDK Pipelines (aws-cdk-lib.pipelines.CodePipeline) would have achieved.  In that during the 'Build/Synth' step of the pipeline CDK assets / artifacts are produced and published to S3.  And from S3 those artifacts are then deployed to CloudFormation.

Is my understanding correct?

I apologize if this is quite clear to any others, but for me it was a struggle.  Why?  I don't have AWS accounts that have the same level of access proposed/needed by the examples.  And it is not in my preferred language of Java.  So I'm not able to confirm that this is indeed the intention/purpose of the project.

**Comment by kaizen3031593** [comment]
Hi @roskelleycj! Thanks for opening this issue. I'll do my best to explain the idea behind this repository.

> Is my understanding correct?

At a high level, the `pipelines` module in `@aws-cdk` is an API for pipelines that is backed by a CodePipeline engine. When developing `pipelines`, we made special care to decouple the generic pipeline API from the CodePipeline "implementation detail". So, this repository, `cdk-pipelines-github`, exists as a proof of concept that we can use the generic pipeline API from `@aws-cdk/pipelines` and attach a different CI/CD engine in the backend (GitHub workflows).

In that sense, we are replacing the CodePipeline backend with GitHub workflows in this repository. So yes, your understanding is correct and the pipeline is fully defined in GitHub Workflows. The workflow will set up a pipeline where your assets will be published to S3, and then the deploy step will consume those artifacts.

> I don't have AWS accounts that have the same level of access proposed/needed by the examples

I would be interested in learning more about this. Are you referring to the OIDC IAM role you need to set up to authenticate your user/repo to AWS? Unfortunately there's no way around that, but this module does provide a [`GitHubActionRole`](https://github.com/cdklabs/cdk-pipelines-github#github-action-role) that aims to make setting it up as smooth as possible.  

> And it is not in my preferred language of Java

It is unfortunately not published to Java yet. Please +1 the issue if you have not already!

**Comment by roskelleycj** [comment]
@kaizen3031593 thank you, again, for the details and confirmation.

I was finally able to acquire an account where the GitHubActionRole was constructed similarly to the [construct](https://github.com/cdklabs/cdk-pipelines-github#github-action-role).  And after a little bit of fiddling w/ the conditions I was able to see a simpler example (single account vs multi-account) work.  Of course it is still in typescript.  😊 but at least I was able to see the entire concept from start to finish work.

It is a novel approach, and really helps if the pipeline is defined in this manner to allow the backend to be different.  Thank you for the help.
